src/main/java/de/hysky/skyblocker/skyblock/dwarven/CrystalsHud.java

```diff
@@ -0,0 +1,129 @@
+     *
+     * @param context DrawContext to draw map to
+     * @param hudX Top left X coordinate of the map
+     * @param hudY Top left Y coordinate of the map
+     */
+    public static void render( DrawContext context, int hudX, int hudY) {
+        //draw map texture
+        context.
+                drawTexture(MAP_TEXTURE,hudX,hudY,0,0,62,62,62,62);
+        //if enabled add waypoint locations to map
+        if (SkyblockerConfigManager.get().locations.dwarvenMines.crystalsHud.showLocations){
+            Map<String,CrystalsWaypoint> ActiveWaypoints=  SkyblockerConfigManager.get().locations.dwarvenMines.crystalsWaypoints.ActiveWaypoints;
+            for (CrystalsWaypoint waypoint : ActiveWaypoints.values()){
+                Color waypointColor = waypoint.category.color;
+                Pair<Integer, Integer> renderPos  = transformLocation(waypoint.pos.getX(),waypoint.pos.getZ());
+                int locationSize  = SkyblockerConfigManager.get().locations.dwarvenMines.crystalsHud.locationSize;
+                if (Arrays.asList(SMALL_LOCATIONS).contains(waypoint.name.getString())){//if small location half the location size
+                    locationSize = locationSize/2;
+                }
+                //fill square of size locationSize around the coordinates of the location
+                context.fill(hudX+renderPos.first()-locationSize/2,hudY+renderPos.second()-locationSize/2,hudX+renderPos.first()+locationSize/2,hudY+renderPos.second()+locationSize/2,waypointColor.getRGB());
+            }
+        }
+        //draw player on map
+        if (client.player == null || client.getNetworkHandler() == null) {
+            return;
+        }
+        //get player location
+        double playerX = client.player.getX();
+        double playerZ = client.player.getZ();
+        double facing = client.player.getYaw();
+        Pair<Integer, Integer> renderPos  = transformLocation(playerX,playerZ);
+        //draw marker on map
+        context.
+                drawTexture(MAP_ICON,hudX+renderPos.first()-2,hudY+renderPos.second()-2,58,2,4,4,128,128);
+
+        //todo add direction and scale (can not work out how to rotate)
+
+    }
+    /**
+     * Converts an X and Z coordinate in the crystal hollow to a X and Y coordinate on the map.
+     *
+     * @param x the world X coordinate
+     * @param z the world Z coordinate
+     * @return the pair of values for x and y
+     */
+    protected static Pair<Integer, Integer> transformLocation(double x, double z){
+        //converts an x and z to a location on the map
+        int transformedX = (int)((x-202)/621 * 62);
+        int transformedY = (int)((z -202)/621 * 62);
+        transformedX = Math.max(0, Math.min(62, transformedX));
+        transformedY = Math.max(0, Math.min(62, transformedY));
```

Comment:
I would replace this with `MathHelper#clamp` (same thing but more straightforward).